File: dsa-problems/ProbabilisticDS/solutions/javascript/PDS-005-misra-gries.js
const readline = require("readline");

function misraGries(stream, k) {
  const counts = new Map();
  
  for (const x of stream) {
    if (counts.has(x)) {
      counts.set(x, counts.get(x) + 1);
    } else if (counts.size < k - 1) {
      counts.set(x, 1);
    } else {
      // Decrement all
      const toRemove = [];
      for (const [key, val] of counts) {
        if (val - 1 === 0) {
          toRemove.push(key);
        } else {
          counts.set(key, val - 1);
        }
      }
      for (const key of toRemove) {
        counts.delete(key);
      }
    }
  }
  
  const res = Array.from(counts.keys()).sort((a, b) => a - b);
  return res;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => {
  const parts = line.trim().split(/\s+/);
  for (const part of parts) {
    if (part !== "") data.push(part);
  }
});
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  const k = parseInt(data[idx++], 10);
  const stream = [];
  for (let i = 0; i < n; i++) stream.push(parseInt(data[idx++], 10));
  const res = misraGries(stream, k);
  console.log(res.join(" "));
});

File: dsa-problems/ProbabilisticDS/solutions/python/PDS-005-misra-gries.py
import sys
from collections import Counter

def misra_gries(stream, k):
    counts = Counter()
    
    for x in stream:
        if x in counts:
            counts[x] += 1
        elif len(counts) < k - 1:
            counts[x] = 1
        else:
            # Decrement all
            to_remove = []
            for key in counts:
                counts[key] -= 1
                if counts[key] == 0:
                    to_remove.append(key)
            for key in to_remove:
                del counts[key]
                
    return sorted(counts.keys())

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        k = int(next(iterator))
        stream = []
        for _ in range(n):
            stream.append(int(next(iterator)))
            
        res = misra_gries(stream, k)
        print(" ".join(str(x) for x in res))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()
